doc: multi-o setup instructions

kyriediculous opened the pull request:
**What does this pull request do? Explain your changes. (required)**
Provide instructions for various ways to start a multi-o setup with a redeemer.

kyriediculous committed:
doc: multi-o setup instructions

yondonfu commented on doc/multi-o.md:
````diff
@@ -0,0 +1,117 @@
+For brevity we'll exclude these flags in the examples by `<...ETH SETUP...>` but these should always be required. If no Ethereum account is available the Livepeer node will create one for you.
+
+```
+livepeer
````
Let's add a backslash after `livepeer` in all of the commands in this doc since it would be required for a valid multi-line bash command.

kyriediculous replied:
fixed in 4131a02b

yondonfu commented on doc/multi-o.md:
````diff
@@ -0,0 +1,117 @@
+```shell
+livepeer
+    <...ETH SETUP...\
+    -orchestrator= true \ 
````
Extra space before true can be removed.

kyriediculous replied:
fixed in 4131a02b

yondonfu commented on doc/multi-o.md:
```diff
@@ -0,0 +1,117 @@
+livepeer
+    -network <mainnet|rinkeby> \
+    -ethUrl <JSON_RPC_PROVIDER> \
+    -ethKeystorePath <PATH_TO_KEY> \
```
`-ethKeystorePath` points to a directory and not to the keystore file itself so this should be updated to something like `<PATH_TO_KEYSTORE_DIR>`.

kyriediculous replied:
fixed in 4131a02b

yondonfu commented on doc/multi-o.md:
```diff
@@ -0,0 +1,117 @@
+
+## Single Orchestrator with Redeemer
+
+This allows an orchestrator running a single on-chain node to use a seperate Ethereum account to redeem winning tickets on-chain and pay for that transaction while the recipient is still orchestrator node's Ethereum address.
```
Suggestion to clarify that the recipient address is the *on-chain registered address*:

"recipient is still orchestrator node's Ethereum address" -> "the ticket recipient is still the on-chain registered Ethereum address".

kyriediculous replied:
fixed in 4131a02b

yondonfu commented on doc/multi-o.md:
```diff
@@ -0,0 +1,117 @@
+# Multi Orchestrator Setup
+
+This document describes the various ways an orchestrator could run multiple nodes and separate concerns. 
```
I think it could be worth clarifying that the nodes would still accept payments on behalf of the same on-chain registered address while each node uses its own address.

kyriediculous replied:
fixed in 4131a02b

yondonfu commented on doc/multi-o.md:
```diff
@@ -0,0 +1,117 @@
+# Multi Orchestrator Setup
+
+This document describes the various ways an orchestrator could run multiple nodes and separate concerns. 
```
For terminology, I suggest we use "orchestrator operator" to refer to the entity running orchestrator nodes. So, this sentence would be updated to: "an orchestrator operator could run multiple nodes...".

j0sh replied:
Could be specific about the node type, eg "an operator can run multiple orchestrator nodes."

I'd also elaborate a bit on the "separate concerns", probably in a follow-on sentence. What concerns are those? Wallet and key management? Blockchain monitoring? Submitting transactions?

kyriediculous replied:
How about we refer to the role as "Orchestrator"  and the nodes as "orchestrator nodes" , wouldn't that be simple yet understandable enough?

yondonfu replied:
"This document describes the various ways an orchestrator could run multiple orchestrator nodes".

vs.

"This document describes the various ways an operator could run multiple orchestrator nodes".

The second option reads a bit smoother to me - I think it makes a clearer distinction between the entity running the nodes and the node type so I have a slight preference for this option.

yondonfu commented on doc/multi-o.md:
```diff
@@ -0,0 +1,117 @@
+livepeer
+    <...ETH SETUP...> \
+    -orchestrator=true -transcoder=true \
+    -initializeRound=true \
```
Orchestrators aren't required to initialize rounds right now which is why the round initializer service is off by default and we don't communicate that orchestrators should be running with this flag right now. What do you think about excluding the round initialization bit from these docs right now and only including it when we're communicating that the flag should be used by orchestrators regardless of what kind of setup they are using?

kyriediculous replied:
I think we should avoid a scenario whereby people feel comfortable omitting this flag, so how about the reverse ? 

Keep it in but mention that livepeer currently 'sponsors' round initialisation but that ideally node operators shouldn't rely on single points of failures like that and make sure they can do everything within their means to run robust infra.

yondonfu replied:
I hear you on wanting orchestrator operators to expect to initialize rounds. But, it is worth noting that we currently do not indicate anywhere in our docs that the `-initializeRound` flag should be used. So, it feels a bit weird to drop in this flag in this specific doc which is focused on multi-O setups when the flag is not mentioned anywhere else. I'd be more inclined to mention the `-initializeRound` flag in a more general purpose doc first along with a description of what it does before including it in this doc.

yondonfu commented on doc/multi-o.md:
```diff
@@ -0,0 +1,117 @@
+livepeer
+    <...ETH SETUP...\
+    -redeemer=true \ 
+    -httpAddr <REDEEMER_HTTP_ADDR (host):port \ 
```
Missing > at the end. Same comment applies for a few of the other flag values in this doc so it'd be good to take a pass through all them to add the missing > at the end.

kyriediculous replied:
fixed in 4131a02b

yondonfu commented on doc/multi-o.md:
````diff
@@ -0,0 +1,117 @@
+```shell
+livepeer
+    <...ETH SETUP...> \
+    -orchestrator=true \
````
Needs `-transcoder`.

kyriediculous replied:
fixed in 4131a02b

yondonfu commented on doc/multi-o.md:
```diff
@@ -0,0 +1,117 @@
+    <...ETH SETUP...> \
+    -orchestrator=true -transcoder=true \
+    -initializeRound=true \
+    -httpAddr <ORCH_HTTP_ADDR (host):port \
```
Might be ok to omit this as the default should be ok.

kyriediculous replied:
Yeah you might be right I think this is covered elsewhere in the docs plenty.

yondonfu replied:
Cool then I think we can remove this (here and in other commands in this doc where it isn't necessary as well) in the interest of focusing on the flags absolute required for these instructions.

j0sh commented on doc/multi-o.md:
```diff
@@ -0,0 +1,117 @@
+
+## Unlocking Account and Connecting to Ethereum
+
+Unlocking your Ethereum account and connecting to a JSON-RPC provider are necessary steps to run the Livepeer node unless it's in standalone transcoder mode. 
```
Maybe use the active voice? "The Livepeer node needs you to unlock your Ethereum account and provide connection information for a JSON-RPC provider, unless the node is in standalone transcoder mode." [1]

Do we have documentation on what connecting to a JSON-RPC provider entails? Maybe link to that, or add a quick blurb if none exists? Also note the slightly modified language here, since the node is the one connecting to the JSON-RPC provider, not the user; the user just needs to provide the connection information.

Also, maybe briefly describe "standalone transcoder" mode, or link if there's a corresponding doc.

j0sh replied:
Also - this section is seems more broadly applicable to all on-chain use cases, rather than multi-O specifically. Probably don't need to refactor the docs right now, but just noting that this part could be extracted into more general-purpose onchain documentation someday.

kyriediculous replied:
fixed in 4131a02

j0sh commented on doc/multi-o.md:
````diff
@@ -0,0 +1,117 @@
+```shell
+livepeer
+    <...ETH SETUP...> \
+    -orchestrator=true -transcoder=true \
````
The `=true` bits are not necessary, and can be omitted here and in all the following examples. Invoking bool flags automatically sets their value to true.

kyriediculous replied:
fixed in 4131a02

j0sh commented on doc/multi-o.md:
```diff
@@ -0,0 +1,117 @@
+
+## Redeemer + RewardService & RoundInitializer + Multiple Orchestrators
+
+This is the "coldest" setup possible for on-chain registered addresses. The keys for the on-chain registered address will be responsible for initialising rounds and calling reward, since reward can not be called on behalf of another address in the current version of the protocol.
```
Maybe split up that last sentence a bit. "The keys for the on-chain registered address will be responsible for calling reward. In the current version of the protocol, `reward` can not be called on behalf of another address."

kyriediculous replied:
removed the second part because we mention it earlier as well, fixed in 4131a02

j0sh commented on doc/multi-o.md:
```diff
@@ -0,0 +1,117 @@
+
+In this setup a node started with the keys for the on-chain registered address will be responsible for all transactions. 
+
+In order to use multiple orchestrators with this setup, the cluster of orchestrator nodes responsible for transcoding would have to be behind a load balancer, **the URI of the load balancer will be the on-chain registered Service URI.**
```
Maybe elaborate on what forms this load balancer could take, since that's not immediately clear from the description here? FWIW, I think DNS is basically the easiest option right now, other than having the broadcaster take an explicit list via webhook or `-orchAddr`. For those interested in building their own LB, it might be worth linking to the networking documentation around GetOrchestrator/OrchestratorInfo (it's unlikely anyone will try to at this stage, but maybe just mention it for completeness).

Also, I think the bolded parts can be split into its own sentence.

kyriediculous replied:
I decided to not include any specifics about that here due to the fact that anything other than simple load balancing between a cluster of nodes within reasonable proximity is still an open area of investigation. We currently only use an explicit list on our broadcasters, which is only an option in a BOT setup. 

I think we can go more into detail once we have a better sense of the specifics here and what would work well for livepeer and what wouldn't ?

j0sh replied:
Yeah, that's reasonable. Should we still mention "load balancing" even though we don't necessarily have concrete guidance around that? Seems like that would invite more questions than it answers. Maybe something like,

"In order to select multiple orchestrator nodes from the same operator, the broadcaster needs to explicitly list each orchestrator, either via the `-orchAddr` or`-orchWebhookUrl` flags. Additional load balancing options are being developed."

DNS should also work reasonably well out-of-the-box, although I don't know if anyone has actually tried it yet. Should we mention that as well?

kyriediculous replied:
While that would work for internal usage or a BOT setup I think it might be good to have these docs also be useful for Orchestrators.  I would avoid mentioning broadcasters here. 

I also feel like the current non-default selection is just an intermediate step that works for us but not necessarily for your average broadcaster node. 

I tried DNS load balancing with NGINX but only for https, it was only a single O but it should work quite the same for multiple O's. 

I added an annotation "e.g. NGINX" but I can do "e.g. DNS Load Balancing" to not be vendor specific if you want.

kyriediculous replied:
replaced with "e.g DNS load balancing"

j0sh commented on doc/multi-o.md:
```diff
@@ -0,0 +1,117 @@
+
+2. Start the RewardService & RoundInitializer 
+
+In the current go-livepeer version (0.5.8) this node type will still run a transcoder server. To avoid this node also transcoding avoid making it accessible through the load balancer of your choice. 
```
This isn't really a documentation issue per se, but just noting that reading through this makes me realize that we really need to loosen up restriction on who can call `reward`, because that would simplify much of the guidance here.

> To avoid this node also transcoding avoid making it accessible through the load balancer of your choice.

An easier way may be to set the ServiceURI to localhost and ensure the box itself is not available to the public?

"To prevent this node from transcoding, set the service URI to localhost and ensure the system is not behind a publicly available IP: `-serviceAddr 127.0.0.1:8935`"

(BTW I just checked this and it seems that we listen to all ipv6 interfaces regardless of what the ipv4 address is, so that needs to be corrected at some point, but should be a non-issue as long as the system itself isn't open to the public, eg NAT'd.)

kyriediculous replied:
>  we really need to loosen up restriction on who can call reward

Yeah we briefly discussed this on discord yesterday. This would be a requirement indeed before we can have complete cold storage for on-chain registered addresses. 

> An easier way may be to set the ServiceURI to localhost and ensure the box itself is not available to the public?

This only overwrites the serviceURI locally and not on-chain, right?

j0sh replied:
> This only overwrites the serviceURI locally and not on-chain, right?

Correct. The user should also ensure the on-chain address points elsewhere, and that the current host not publicly available.

kyriediculous replied:
This should no longer be necessary with the latest changes in #1547

yondonfu commented on doc/multi-o.md:
```diff
@@ -0,0 +1,123 @@
+
+## Single Orchestrator with Redeemer
+
+This setup allows an Orchestrator to use a seperate Ethereum account for redeeming winning tickets on-chain and paying for that transaction while the ticket recipient is still Orcherstrator's on-chain registered Ethereum address.
```
"still Orchestrator's" -> "still the Orchestrator's..."

yondonfu commented on doc/multi-o.md:
````diff
@@ -0,0 +1,123 @@
+1. Start the Redeemer
+
+```shell
+livepeer
````
Missing backslash for this commands and the ones after it.

yondonfu commented on doc/multi-o.md:
````diff
@@ -0,0 +1,123 @@
+```shell
+livepeer
+    <...ETH SETUP...> \
+    -orchestrator -trancsoder \
````
With the recent changes in #1547 this command can instead be:

```golang
livepeer \
    <...ETH SETUP...> \
    -initializeRound \
    -ethOrchAddr <ORCHESTRATOR_ON_CHAIN_ETH_ADDR> \
```

We can add a note that the address specified via `-ethAcctAddr` should be the same as the address specified by `-ethOrchAddr`.

kyriediculous replied:
I don't think this is possible because on must always supply a node type, orchestrator or redeemer in this case. Even with the changes in #1547

kyriediculous replied:
fixed in 86cebba

yondonfu commented on doc/multi-o.md:
```diff
@@ -0,0 +1,123 @@
+    <...ETH SETUP...> \
+    -redeemer \ 
+    -httpAddr <REDEEMER_HTTP_ADDR (host):port> \ 
+    -ethOrchAddr <ORCHESTRATOR_ON_CHAIN_ETH_ADDR>
```
With the recent changes in #1547 we can add a note that the address specified via `-ethAcctAddr` should be different than the address specified by `-ethOrchAddr` (or else the reward service will be started with this command).

kyriediculous replied:
We already have a note that states the address should be that of the rewardservice, but i'll elaborate on it.

kyriediculous replied:
fixed in 86cebba

yondonfu commented on doc/multi-o.md:
````diff
@@ -0,0 +1,119 @@
+    -pricePerUnit <PRICE (wei/pixel if '-pixelsPerUnit' is not set)>
+```
+
+## Redeemer + RewardService  + Multiple Orchestrators
````
Let's remove the extra space after "RewardService".

kyriediculous replied:
fixed in daacf97

yondonfu commented on doc/multi-o.md:
````diff
@@ -0,0 +1,115 @@
+```
+
+
+## Blockchain Service Node with (Multiple) Orchestrator(s) node(s)
````
The (s) after Orchestrator can be removed if it is followed by "node(s)".

yondonfu reviewed (commented):
Just one comment - feel free to address it and rebase

kyriediculous commented:
addressed the comment and rebased

On Wed, Jul 1, 2020 at 8:37 PM Yondon Fu <notifications@github.com> wrote:

> *@yondonfu* commented on this pull request.
>
> Just one comment - feel free to address it and rebase
> ------------------------------
>
> In doc/multi-o.md
> <https://github.com/livepeer/go-livepeer/pull/1548#discussion_r448546972>:
>
> > +    -ethOrchAddr <ORCHESTRATOR_ON_CHAIN_ETH_ADDR (also the recipient address)>
> +```
> +
> +2. Start the Orchestrator node
> +
> +
> +```shell
> +livepeer \
> +    <...ETH SETUP...> \
> +    -orchestrator -transcoder \
> +    -redeemerAddr <REDEEMER_HTTP_ADDR> \
> +    -pricePerUnit <PRICE (wei/pixel if '-pixelsPerUnit' is not set)>
> +```
> +
> +
> +## Blockchain Service Node with (Multiple) Orchestrator(s) node(s)
>
> The (s) after Orchestrator can be removed if it is followed by "node(s)".
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/livepeer/go-livepeer/pull/1548#pullrequestreview-441108610>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFJZZWWVNYE4XJ2MUTTOVT3RZN645ANCNFSM4OI26MZQ>
> .
>

yondonfu reviewed (approved):
LGTM
